sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala

```diff
@@ -202,6 +209,27 @@ private[engine] final class Preprocessor(
       } yield m.updated(pkgName, pkgId)
     )
 
+  def buildGlobalKey(
+      templateId: Ref.TypeConName,
+      contractKey: Value,
+  ): Result[GlobalKey] = {
+    safelyRun(pullPackage(Seq(templateId))) {
+      val ckType = handleLookup(pkgInterface.lookupTemplateKey(templateId)).typ
+      val languageVersion =
+        handleLookup(pkgInterface.lookupPackageLanguageVersion(templateId.packageId))
+      val sValue = valueTranslator.unsafeTranslateValue(
+        ty = ckType,
+        upgradable = languageVersion >= LanguageVersion.Features.smartContractUpgrade,
+        value = contractKey,
+      )
+      speedy.Speedy.Machine
+        .globalKey(pkgInterface, templateId, sValue)
+        .getOrElse(
+          throw Error.Preprocessing.ContractIdInContractKey(contractKey)
+        )
```

Comment:
This looks very similar to https://github.com/digital-asset/daml/blob/f14fc8d20e7d0b569483d9ae43c707d7cf1c85ee/sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala#L342-L366

Can you reuse the method implementation?